src/wp-includes/class-wp-block-bindings-source.php

```diff
@@ -60,6 +68,9 @@ public function __construct( string $name, array $source_properties ) {
 		$this->name               = $name;
 		$this->label              = $source_properties['label'];
 		$this->get_value_callback = $source_properties['get_value_callback'];
+		if ( isset( $source_properties['uses_context'] ) ) {
```

Comment:
I think that we should move this `isset()` check next to the [other checks](https://github.com/SantosGuillamot/wordpress-develop/blob/c4fb33fdaee7388c6246b3857d10696666f620ab/src/wp-includes/class-wp-block-bindings-registry.php#L123-L141)
in `WP_Block_Bindings_Registry`.

Reply:
And let's cover it with a unit test as well 
🙂. (Should be almost a copy-paste of the
existing ones)

Reply:
My idea is to make the `uses_context` optional because some sources won't need to use it. Do you think it shouldn't be that case? It would make sense to add a check that, if it exists, it is an array though.

I'll work on unit tests as well 🙂

Reply:
Yes, it should be optional and default to `null`.

Reply:
Right, that makes sense! It should be optional indeed 🙂

Reply:
I've added the check [here](https://github.com/WordPress/wordpress-develop/pull/6079/files#diff-3fd1b8d9e55d4fe09b5f48002d360d3ba9d338d6f33ef3959828424a590b4947R154-R162).

I'll check the unit tests.

Reply:
As explained here, I added a few test to ensure it works as expected: [link](https://github.com/WordPress/wordpress-develop/pull/6079#issuecomment-1945789396).

Reply:
There are two options in the constructor:

- keep adding `isset` checks for the settings that are optional
- ensure that we pass only allowed settings in `$source_properties` in the registry and add `foreach` loop similar to:
  https://github.com/WordPress/wordpress-develop/blob/7b24083859c37d8ac75b9549f458bc389610eaef/src/wp-includes/class-wp-block-type.php#L576C3-L578C4

Reply:
If I understood you correctly, I believe I prefer the second option. I made [this commit](https://github.com/WordPress/wordpress-develop/pull/6079/commits/07a7317882354bb68304a34a6c929c12b45d424f) to illustrate how I understood it. We can always revert it.

_PD: I included unnecessary code I was using for testing, I just removed it in other commit_ 😅

Reply:
Yes, it's exactly how it could work 👍🏻 

It would be nice to add a unit test for the registry that asserts that it errors when an unknown property is provided.

---

At the same time, we are closing the source object for an extension, which I'm a bit unsure about, given we might need to change during beta if we figure out it prevents iterations in the Gutenberg plugin. We also miss the following annotation on the class if we decide to go that path:

https://github.com/WordPress/wordpress-develop/blob/7b24083859c37d8ac75b9549f458bc389610eaef/src/wp-includes/class-wp-block-type.php#L17

Reply:
I am unsure about restricting it as well. I can expect we might want to increase it in Gutenberg after external developers create their own sources the same way we need now to include `uses_context`.

Reply:
On the other hand, if we allow source to use their own properties, could we face backwards-compatibility issues in the future?

Imagine a source is using a custom property `format`. And in the future, we want to include it as part of the core properties and we use it somehow as we are doing with `uses_context`. The source that had a custom property with the same name wouldn't be working as expected, right?

Reply:
I'll keep thinking about it as I don't have an ultimate answer for now as it's hard to tell what will be needed here.